### PR TITLE
standardization in plugin

### PR DESCRIPTION
**Pull request by vinpel**
## Contribution type
Bugfix, feature : a lot of modification, need to be tested and reviewed

### Added

  - **"priority_path"**, new accepted value : **"binary_path"**, "local" is  now accepted too
  - global :
    -  **binary_path** to manualy set a directory to find plugins binary (interpolated directory %%)

  - per plugin :
    - **"binary_path"** : set the path to find the finary, overwrite the global **"binary_path"** option if set
    - **"binary_name"** : change the default name of binary, array accepted
    - **"priority_path"** : can be defined per plugin
     
### Changed

- Per plugin:
    - **ignore** : always define path  from root without ./ ,it will be merge with the global **ignore** directory list
- All plugins who previously used "path" or "directory" option now use a common **"directory"**
  it's the working directory of the plugin

```yaml
test:
  php_cpd:
    directory: src/agents/
    binary_path: /custom/binary
    binary_name: php_cpd-1-2
    allow_failures: true
    ignore:
      - src/components/method/
      - src/agents/config/models
      - tests/
```
In our exemple, only  "src/agents/config/models" is inside directory "src/agents" and we pass only the subdirectory :
```shell
Args: ["\/app\/php-censor\/vendor\/bin\/phpcpd --log-pmd \"%s\" %s \"%s\"","\/tmp\/php_cpd_KrEPQo"," --exclude config\/models","\/app\/php-censor\/runtime\/builds\/1\/115_fc99faf0\/src\/agents\/"]
Command: /app/php-censor/vendor/bin/phpcpd --log-pmd "/tmp/php_cpd_KrEPQo"  --exclude config/models "/app/php-censor/runtime/builds/1/115_fc99faf0/src/agents/"
```
it allow plugin to correctly ignore directory

__Need review :__

This modification has been removed, need to fix a unit test before and mock Config::singleton
 added %DATABASE_TEST_HOST% keywork, now write of database, in my  case host can be different, used like this :

```yaml
shell:
- "sed -i -e 's,mysql:host=db;dbname=database_test,mysql:host=%DATABASE_TEST_HOST%;dbname=TEST_DB_BUILD_%BUILD%,g' %BUILD_PATH%src/config/db_mysql_test.php"
```

### CODE REVIEW

**Comment by vinpel** [comment]
i will look at "priority_path" things

**Comment by corpsee** [comment]
@vinpel Thank you! I will try to review request as soon as possible.

**Comment by corpsee** [comment]
> i will look at "priority_path" things

I mean It is simpler to add common option in the `Plugin` class than add similar options to all plugins.

File: docs/en/plugins/codeception.md [hunk 1]
```diff
@@ -15,9 +15,9 @@ configuration file paths. By default this is called `codeception.yml` and will b
 made on line 132 regarding the value of `--xml` being the next argument which will not be correct if the user provides 
 arguments using this config param, you must specify `report.xml` before any user input arguments to satisfy the report 
 processing on line 146.
-
-* **path** - Optional - The path from the root of your project to the root of the codeception `_output` directory
-
+* **path** - Deprecated - use output_path
```
Comment (corpsee) [inline docs/en/plugins/codeception.md]: @vinpel Could you please also add `@deprecated` phpdoc tag to the code where used the option?

**Comment by vinpel** [comment]
> 
> 
> > i will look at "priority_path" things
> 
> I mean It is simpler to add common option in the `Plugin` class than add similar options to all plugins.

it depend, certains plugin didn't interprete the same way directory or didn't have the option

**Comment by corpsee** [comment]
@vinpel From my point of view if plugin doesn't have binary file or any own options it ignore the option and  do nothing (`priority_path` works same for plugins without binary).

It will simplier to support in the future than many similar options in each plugin. I don't tell about `directory` option, I tell only about `executable` option.

**Comment by corpsee** [comment]
@vinpel Anyway I will approve request but I want than you will think about my proposal.

**Comment by corpsee** [comment]
@vinpel [Tests failed](https://travis-ci.org/php-censor/php-censor/builds/470555657?utm_source=github_status&utm_medium=notification). Could you fix it?

**Comment by vinpel** [comment]
I have fixed the test, updated the desciption
 (you may notice i have changed the options name to be more consistent)

**Comment by corpsee** [comment]
@vinpel Thanks! Great job! :+1:

**Pull request merged**